### PR TITLE
[symint] add support for negative integers

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #77795
* #77914
* __->__ #77913

The bit packing scheme is described in the comments.

### CODE REVIEW

File: c10/core/SymInt.h [hunk 1]
```diff
@@ -63,7 +58,22 @@ class C10_API SymInt {
...
  private:
-  const static int64_t SYM_TAG_MASK = 1LL << 63;
+  // Constraints on the internal representation:
+  // - Should represent positive and negative ints
+  // - No conversion necessary for operations on ints.
+  // - We reserve some values to act as indices into our sym int table.
+  //
+  // So, the scheme is to reserve large negative numbers:
+  // - 0b11... means we are a negative int (following two's complement)
+  // - 0b00... means we are a positive int (following two's complement)
```
Comment (Krovatkin) [inline c10/core/SymInt.h:68]: nit: INT64_MAX is 0b0111111...., so positive numbers are both 0b00xxx and 0b01xxx?

File: c10/core/SymInt.h [hunk 2]
```diff
@@ -63,7 +58,22 @@ class C10_API SymInt {
...
+  // - 0b10... means we are index into the sym table. This means that
+  //           [-2^63, -2^62-1] are not representable as ints.
+  static constexpr uint64_t MASK = 1LL << 63 | 1LL << 62;
```
Comment (Krovatkin) [inline c10/core/SymInt.h:71]: maybe, ULL?

File: c10/test/core/SymInt_test.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  EXPECT_FALSE(i.is_symbolic());
+  EXPECT_EQ(i.data(), value);
+}
+
+TEST(SymIntTest, ConcreteInts) {
+  check(INT64_MAX);
+  check(0);
+  check(-1);
+  // This is 2^62, which is the most negative number we can support.
+  check(-4611686018427387904);
```
Comment (Krovatkin) [inline c10/test/core/SymInt_test.cpp:19]: nit: can we stick the LL suffix to make sure a compiler won't discard the higher 32bits of this literal?

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: c10/core/SymInt.h (hunk 1 above)
Comment (suo) [reply c10/core/SymInt.h]: rewrote the comment!

**Pull request closed without merging**